Consistent handling of out-of-flow children in MathML layout functions

icesfont opened the pull request:
#### 7c0a9d18a4e012bcfba7939a6bf697e50ad079b1
<pre>
Consistent handling of out-of-flow children in MathML layout functions

<a href="https://bugs.webkit.org/show_bug.cgi?id=179739">https://bugs.webkit.org/show_bug.cgi?id=179739</a>

Reviewed by Frédéric Wang.

According to the spec [1], out-of-flow children should not affect the rendering of MathML elements. Ensure that out-of-flow children are still laid out and rendered, but that layout/validity logic only concerns in-flow children.

WebKit does not follow the spec specifically here ([2]), causing floating children to be OOF:

&gt; The float property does not create floating of elements whose parent&apos;s computed display value is block math or inline math, and does not take them out-of-flow.

Cases in the WPT test in [3] used to pass regardless, since layout functions for MathML elements did not consider flow. Because they now *do* consider flow, the float property is no longer ignored in layout (as it causes elements to be OOF) and cases in the WPT test fail.

See webkit.org/b/279978.

[1] <a href="https://w3c.github.io/mathml-core/">https://w3c.github.io/mathml-core/</a>
[2] <a href="https://w3c.github.io/mathml-core/#css-styling">https://w3c.github.io/mathml-core/#css-styling</a>
[3] <a href="https://github.com/web-platform-tests/wpt/blob/4d67f35a4dcc96bc40de00102a150d925d9fca61/mathml/relations/css-styling/floats/not-floating-001.html">https://github.com/web-platform-tests/wpt/blob/4d67f35a4dcc96bc40de00102a150d925d9fca61/mathml/relations/css-styling/floats/not-floating-001.html</a>

Canonical link: <a href="https://commits.webkit.org/284225@main">https://commits.webkit.org/284225@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/74665ac37a81ec86d0f5d08d4b1ee25633d41a66

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68776 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48168 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21435 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72846 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19921 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70893 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55964 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19737 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54816 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13250 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71843 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44009 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59382 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35283 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40674 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16808 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18279 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62623 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17156 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74540 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12747 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16403 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62304 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12787 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59464 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62341 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15267 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10297 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3908 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43969 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45043 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46237 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44785 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLPadded.cpp:
```diff
@@ -122,8 +123,9 @@ void RenderMathMLPadded::layoutBlock(bool relayoutChildren, LayoutUnit)
 std::optional<LayoutUnit> RenderMathMLPadded::firstLineBaseline() const
 {
     // We try and calculate the baseline from the position of the first child.
+    // TODO: needs assert to make sure that firstInFlowChildBox() exists?
```
Why? If there it does not then we go the else clause below (`ascent = mpaddedHeight(0)`)

icesfont replied:
Oops, yes, of course

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLBlock.cpp:
```diff
@@ -275,6 +275,16 @@ void RenderMathMLBlock::styleDidChange(StyleDifference diff, const RenderStyle*
+void RenderMathMLBlock::insertPositionedChildrenInContainingBlock()
+{
+    for (auto* child = firstChildBox(); child; child = child->nextSiblingBox()) {
+        if (!child->isInFlow()) {
```
mmh, I believe `child->isOutOfFlowPositioned()` from your previous patches was correct.  The name `insertPositionedObject` really seems it expects oof positioned, not floats ; and the layout of floats is probably handled differently anyway...

But still if you keep `child->isOutOfFlowPositioned()` here and only layout the in-flow children in `layoutBlock` methods, I believe floats won't be laid out,  so `not-floating-001.html` will crash with an assertion failure. You can see that in your previous patches: https://ews-build.s3-us-west-2.amazonaws.com/macOS-Sonoma-Debug-WK2-Tests-EWS/c1f65df6-11348/imported/w3c/web-platform-tests/mathml/relations/css-styling/floats/not-floating-001-crash-log.txt

In https://w3c.github.io/mathml-core/#css-styling, MathML core contains the same text as in CSS grid & flexbox specs:

> The [float](https://www.w3.org/TR/CSS22/#propdef-float) property does not create floating of elements whose parent's computed display value is block math or inline math, and does not take them out-of-flow.

So I believe the right way to make not-floating-001.html pass is to replicate whatever was done for grid/flexbox.

fred-wang replied:
As I see, WebKit currently forces the computed value of float to none for flex and grid displays:

https://searchfox.org/wubkat/rev/b36cbce69fddb7da33823f316bd8ead5bebee970/Source/WebCore/style/StyleAdjuster.cpp#492

(This behavior was changed in Chromium [here](https://chromium-review.googlesource.com/c/chromium/src/+/2157889) so that the computed value remains unchanged:  and the behavior extended to `display: math` [here](https://chromium-review.googlesource.com/c/chromium/src/+/4055563))

In order to fix the crash with floats not laid out, I believe you can just do the same for MathML. However, implementing https://bugs.webkit.org/show_bug.cgi?id=278533 is probably too much and I'm not sure it's ok to access the m_element's parent in the style adjuster.

So I wonder whether we could just have a function that performs dummy layout for floats (with some FIXME comment).

icesfont replied:
I see, thanks -- does the current change work?

fred-wang replied:
Which change? I don't see any new commit...

My point is that (1) insertPositionedObject probably shouldn't be called for floats but (2) we should still need layout floats somehow until we are able to have the flex/grid behavior.

icesfont replied:
Right, I was still building it sorry! The change is there now

fred-wang reviewed (changes requested):
Please update the test expectations for the tests that are now passing.

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLBlock.cpp:
```diff
@@ -275,6 +275,25 @@ void RenderMathMLBlock::styleDidChange(StyleDifference diff, const RenderStyle*
+        // Since this leaves floats as neither positioned nor in-flow, this performs dummy layout to ensure that they are laid out at all.
+        // FIXME: Correctly handle floating children of computed display value `math` elements, and remove this.
+        // [1]: https://w3c.github.io/mathml-core/#css-styling
+        // [2]: https://bugs.webkit.org/show_bug.cgi?id=278533
```
I don't know if adding [1] [2] references are quite common in WebKit code. In any case, maybe this could just be

```
// Perform dummy layout for floating children.
// FIXME: Per the spec, there should be no floating children inside MathML renderers. See webkit.org/b/278533
```

and all relevant information moved directly in the bug.

icesfont replied:
Done

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLFraction.cpp:
```diff
@@ -259,10 +259,7 @@ void RenderMathMLFraction::layoutBlock(bool relayoutChildren, LayoutUnit)
-        numerator().containingBlock()->insertPositionedObject(numerator());
-    if (denominator().isOutOfFlowPositioned())
-        denominator().containingBlock()->insertPositionedObject(denominator());
+    handlePositionedAndFloatingChildren();
```
I would move this generic thing just before the fraction-specific layout starts (before the `numerator().layoutIfNeeded()` call).

icesfont replied:
Done

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLOperator.cpp:
```diff
@@ -227,6 +227,8 @@ void RenderMathMLOperator::layoutBlock(bool relayoutChildren, LayoutUnit pageLog
     LayoutUnit trailingSpaceValue = trailingSpace();
 
+    handlePositionedAndFloatingChildren();
+
```
ditto, just before  `LayoutUnit leadingSpaceValue = leadingSpace();`

icesfont replied:
Done

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLRow.cpp:
```diff
@@ -78,6 +78,8 @@ static RenderMathMLOperator* toVerticalStretchyOperator(RenderBox* box)
+
     // First calculate stretch ascent and descent.
     LayoutUnit stretchAscent, stretchDescent;
     for (auto* child = firstChildBox(); child; child = child->nextSiblingBox()) {
```
Why aren't you browsing only in-flow children in this loop and in the loop below? and removing the `isOutOfFlowPositioned()` checks?

icesfont replied:
Oops, yep

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLRow.cpp:
```diff
@@ -78,6 +78,8 @@ static RenderMathMLOperator* toVerticalStretchyOperator(RenderBox* box)
 void RenderMathMLRow::stretchVerticalOperatorsAndLayoutChildren()
 {
+    handlePositionedAndFloatingChildren();
+
```
Since you are doing that, why aren't you browsing only in-flow children in the two loops below? and removing the `isOutOfFlowPositioned()` checks?

I think this would only be a behavior change for floats, but I understand that's expected for now.

icesfont replied:
Yep -- done

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLRow.cpp:
It seems I cannot comment on specific lines but there are many places in this file where you still browse the whole list of children and use `isOutOfFlowPositioned()`.

See for example:

https://searchfox.org/wubkat/search?q=isOutOfFlowPositioned&path=Source%2FWebCore%2Frendering%2Fmathml&case=false&regexp=false

https://searchfox.org/wubkat/search?q=child+%3D+firstChildBox%28%29%3B&path=Source%2FWebCore%2Frendering%2Fmathml&case=false&regexp=false

icesfont replied:
Thanks, they should be changed now

fred-wang replied:
Sounds good. Maybe the functions can be renamed to replace "Children" and "RowItems" with "InFlowChildren"? (probably in a follow-up patch).

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLToken.cpp:
```diff
@@ -573,10 +573,7 @@ void RenderMathMLToken::layoutBlock(bool relayoutChildren, LayoutUnit pageLogica
-        if (box.isOutOfFlowPositioned())
-            box.containingBlock()->insertPositionedObject(box);
-    }
+    handlePositionedAndFloatingChildren();
```
Other methods do that after the "!relayoutChildren && simplifiedLayout()" early return. Have you checked what non-MathML classes do?

https://searchfox.org/wubkat/search?q=%21relayoutChildren+%26%26+simplifiedLayout%28%29&path=Source%2FWebCore%2Frendering&case=false&regexp=false

It seems at least for flexbox and grid, the insertPositionedObject calls are done *after* the early return.

icesfont replied:
Since this was the previous behaviour will this change anything? I don't think so (mimics what is done for other elements, so it should be fine) but want to make sure

fred-wang replied:
Good question. I believe the only difference is that  the OOF children are inserted yet in the containing block when `simplifiedLayout()` is executed. But we don't care since it's up to the containing block (which is an ancestor) to take care of the layout of these children. `simplifiedLayout()` calls `layoutPositionedObjects()` for the objects from its own positioned objects map (which were inserted by some descendants).

For this particular case, I guess you checked the test added in https://commits.webkit.org/262449@main does not crash in debug?

fred-wang replied:
mmh, actually the containingBlock() will likely return the mathml renderer itself rather than an ancestor, we will need to think about it.

fred-wang replied:
For future reference, Harry and I just had a meeting where we talked about it.

Probably we need to do the insertion of positioned children before the `simplifiedLayout()`, especially for MathML where `child->containingBlock()` is the object itself. Otherwise https://searchfox.org/wubkat/rev/b36cbce69fddb7da33823f316bd8ead5bebee970/Source/WebCore/rendering/RenderBlock.cpp#847 won't layout the inserted positioned children and nobody else we do it for us. (Perhaps for grid/flex, the containing block is always an ancestor so we don't have this issue).

For this edge case to happen, a test case needs to insert a positioned absolute object and do some extra changes that would trigger a `simplifiedLayout()`. Not sure whether that's already cover and how easy it would be to write one, though.

Doing the layout of floats before `simplifiedLayout()` could have a perf impact, so probably better to continue doing it after the `simplifiedLayout()` call, and so split `handlePositionedAndFloatingChildren()`. That's a pity to have multiple loops over the children, but hopefully won't have big impact on performance because each loop is skipping  "irrelevant" children.

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLUnderOver.cpp:
```diff
@@ -296,10 +296,7 @@ void RenderMathMLUnderOver::layoutBlock(bool relayoutChildren, LayoutUnit pageLo
-        if (box.isOutOfFlowPositioned())
-            box.containingBlock()->insertPositionedObject(box);
-    }
+    handlePositionedAndFloatingChildren();
```
ditto, this should probably be after the "!relayoutChildren && simplifiedLayout()" early return.

fred-wang reviewed (changes requested):
I think the commit message should explain a bit about the issue with floats and why some new tests are failing for now.

Also why does https://github.com/WebKit/WebKit/pull/31676#issue-2445910169 show multiple commits ?

fred-wang reviewed (changes requested):
I believe RenderMathMLMath is missing  the `insertPositionedChildrenIntoContainingBlock()` and `layoutFloatingChildren()` calls, which likely explains why the mfrac is not laid out here: https://ews-build.s3-us-west-2.amazonaws.com/macOS-Sonoma-Debug-WK2-Tests-EWS/79d85e7b-16118-rerun/mathml/out-of-flow-fraction-content-crash-crash-log.txt

(please check we do that for all RenderMathML* classes)

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLRoot.cpp:
```diff
@@ -200,11 +202,14 @@ void RenderMathMLRoot::layoutBlock(bool relayoutChildren, LayoutUnit)
     LayoutUnit baseAscent, baseDescent;
     recomputeLogicalWidth();
     computeAndSetBlockDirectionMarginsOfChildren();
+
```
nit: please remove this whitespace change

icesfont replied:
Fixed

fred-wang commented on Source/WebCore/rendering/mathml/RenderMathMLScripts.cpp:
```diff
@@ -74,9 +74,9 @@ RenderMathMLOperator* RenderMathMLScripts::unembellishedOperator() const
 std::optional<RenderMathMLScripts::ReferenceChildren> RenderMathMLScripts::validateAndGetReferenceChildren() const
 {
-    // All scripted elements must have at least one child.
+    // All scripted elements must have at least one in-flow  child.
```
there should be only one space after in-flow

icesfont replied:
Fixed

fred-wang reviewed (commented):
Looks good to me! Please update the expectations of mathml/relations/css-styling/ignored-properties-001 on mac/ios and I think the patch should be ready.

icesfont committed:
Consistent handling of out-of-flow children in MathML layout functions

https://bugs.webkit.org/show_bug.cgi?id=179739

Reviewed by Frédéric Wang.

According to the spec [1], out-of-flow children should not affect the rendering of MathML elements. Ensure that out-of-flow children are still laid out and rendered, but that layout/validity logic only concerns in-flow children.

WebKit does not follow the spec specifically here ([2]), causing floating children to be OOF:

> The float property does not create floating of elements whose parent's computed display value is block math or inline math, and does not take them out-of-flow.

Cases in the WPT test in [3] used to pass regardless, since layout functions for MathML elements did not consider flow. Because they now *do* consider flow, the float property is no longer ignored in layout (as it causes elements to be OOF) and cases in the WPT test fail.

See webkit.org/b/279978.

[1] https://w3c.github.io/mathml-core/
[2] https://w3c.github.io/mathml-core/#css-styling
[3] https://github.com/web-platform-tests/wpt/blob/4d67f35a4dcc96bc40de00102a150d925d9fca61/mathml/relations/css-styling/floats/not-floating-001.html

Canonical link: https://commits.webkit.org/284225@main
